OpInfo: Add test that sample_inputs_func returns a generator

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #84567

This also includes a small list exception for single element lists since none of the memory usage or performance implications of lists apply there.

peterbell10 committed:
OpInfo: Add test that sample_inputs_func returns a generator

[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Add test that sample_inputs_func returns a generator"




[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Add test that sample_inputs_func returns a generator"




[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Add test that sample_inputs_func returns a generator"




[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Add test that sample_inputs_func returns a generator"




[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Add test that sample_inputs_func returns a generator"




[ghstack-poisoned]

peterbell10 committed:
Rebase on "OpInfo: Add test that sample_inputs_func returns a generator"




[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Add test that sample_inputs_func returns a generator"




[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Add test that sample_inputs_func returns a generator"




[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Add test that sample_inputs_func returns a generator"




[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Add test that sample_inputs_func returns a generator"




[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Add test that sample_inputs_func returns a generator"




[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Add test that sample_inputs_func returns a generator"


This does include a small list exception for less than 4 elements,
but that will be tightened to 1 element in a follow-up PR.

[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Add test that sample_inputs_func returns a generator"


This does include a small list exception for less than 4 elements,
but that will be tightened to 1 element in a follow-up PR.

[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Add test that sample_inputs_func returns a generator"


This does include a small list exception for less than 4 elements,
but that will be tightened to 1 element in a follow-up PR.

[ghstack-poisoned]

peterbell10 committed:
Rebase on "OpInfo: Add test that sample_inputs_func returns a generator"


This does include a small list exception for less than 4 elements,
but that will be tightened to 1 element in a follow-up PR.

[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Add test that sample_inputs_func returns a generator"


This does include a small list exception for less than 4 elements,
but that will be tightened to 1 element in a follow-up PR.

[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Add test that sample_inputs_func returns a generator"


This does include a small list exception for less than 4 elements,
but that will be tightened to 1 element in a follow-up PR.

[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Add test that sample_inputs_func returns a generator"


This does include a small list exception for less than 4 elements,
but that will be tightened to 1 element in a follow-up PR.

[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Add test that sample_inputs_func returns a generator"


This does include a small list exception for less than 4 elements,
but that will be tightened to 1 element in a follow-up PR.

[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Add test that sample_inputs_func returns a generator"


This does include a small list exception for less than 4 elements,
but that will be tightened to 1 element in a follow-up PR.

[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Add test that sample_inputs_func returns a generator"


This does include a small list exception for less than 4 elements,
but that will be tightened to 1 element in a follow-up PR.

[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Add test that sample_inputs_func returns a generator"


This also includes a small list exception for single element lists.

[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Add test that sample_inputs_func returns a generator"


This also includes a small list exception for single element lists.

[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Add test that sample_inputs_func returns a generator"


This also includes a small list exception for single element lists.

[ghstack-poisoned]

peterbell10 committed:
Rebase on "OpInfo: Add test that sample_inputs_func returns a generator"


This also includes a small list exception for single element lists.

[ghstack-poisoned]

peterbell10 committed:
Rebase on "OpInfo: Add test that sample_inputs_func returns a generator"


This also includes a small list exception for single element lists.

[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Add test that sample_inputs_func returns a generator"


This also includes a small list exception for single element lists.

[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Add test that sample_inputs_func returns a generator"


This also includes a small list exception for single element lists.

[ghstack-poisoned]

peterbell10 committed:
Rebase on "OpInfo: Add test that sample_inputs_func returns a generator"


This also includes a small list exception for single element lists.

[ghstack-poisoned]

peterbell10 committed:
Update on "OpInfo: Add test that sample_inputs_func returns a generator"


This also includes a small list exception for single element lists.

[ghstack-poisoned]

mruberry commented on test/test_testing.py:
```diff
@@ -1881,5 +1881,35 @@ def test_sample_input_metadata(self) -> None:
         self.assertEqual(s2.name, "foo")
 
 
+class TestOpInfoSampleFunctions(TestCase):
```
Add a comment describing this class

mruberry commented on test/test_testing.py:
```diff
@@ -1881,5 +1881,35 @@ def test_sample_input_metadata(self) -> None:
+        else:
+            self.assertIsInstance(item, Generator)
+
+    @ops(op_db, dtypes=OpDTypes.any_one)
```
These tests make a lot of sense (comments describing each one, please!) 

Should we consider adding these checks to the tests in `test_ops.py` that exercise these generators, instead? That's not conceptually ideal because then those tests are testing multiple things (they probably already are), but it would reduce duplicative work in the test suite, which does a lot already. Or are these new tests so fast we shouldn't be concerned about their speed?

An example of a similar case is the test for "input" tensors having no grad history, which is in the noncontiguous samples test:

https://github.com/pytorch/pytorch/blob/c27a5171b80b126917ef4435e232055415fcf617/test/test_ops.py#L503

cc @lezcano

lezcano replied:
A great thing to have would be a test that tests OpInfo samples themselves, and have all these tests grouped there. Until then, if this test is fast enough, it may be fine.

peterbell10 replied:
On my machine 
```
time pytest test_testing.py -k TestOpInfoSampleFunctions
```
takes less than 10 seconds even on a build with CUDA. So I'm not particularly concerned.

On the other hand if we combine these checks into an existing test, then the test is more likely to be skipped for no good reason.

lezcano replied:
SGTM

peterbell10 committed:
Update on "OpInfo: Add test that sample_inputs_func returns a generator"


This also includes a small list exception for single element lists since none of the memory usage or performance implications of lists apply there.

[ghstack-poisoned]

lezcano reviewed (approved):
Good stuff.

mruberry reviewed (approved):
Timing looks good, but see notes requesting additional comments and the lint failure may be real!

peterbell10 committed:
Update on "OpInfo: Add test that sample_inputs_func returns a generator"


This also includes a small list exception for single element lists since none of the memory usage or performance implications of lists apply there.

[ghstack-poisoned]

peterbell10 committed:
Fix merge conflicts on "OpInfo: Add test that sample_inputs_func returns a generator"


This also includes a small list exception for single element lists since none of the memory usage or performance implications of lists apply there.

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge

peterbell10 committed:
Update on "OpInfo: Add test that sample_inputs_func returns a generator"


This also includes a small list exception for single element lists since none of the memory usage or performance implications of lists apply there.

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge
